book/src/design/circuit/gadgets/sinsemilla/commit-ivk.md

```diff
@@ -0,0 +1,229 @@
+
+    a) $0 \leq a < 2^{130}$. This is expressed in the custom gate as
+       $$b_1 \cdot z_{a,13} = 0,$$
+       where $z_{a,13}$ is the index-13 running sum output by $\SinsemillaHash(a).$
+
+    b) $0 \leq a + 2^{130} - t_\mathbb{P} < 2^{130}$. To check this, we decompose
+       $a' = a + 2^{130} - t_\mathbb{P}$ into thirteen 10-bit words (little-endian) using
+       a running sum $z_{a'}$, looking up each word in a $10$-bit lookup table. We then
+       enforce in the custom gate that
+       $$b_1 \cdot z_{a',13} = 0.$$
+
+$$
+\begin{array}{|c|l|}
+\hline
+\text{Degree} & \text{Constraint} \\\hline
+3 & q_\CommitIvk \cdot b_1 \cdot b_0 = 0 \\\hline
+3 & q_\CommitIvk \cdot b_1 \cdot z_{a,13} = 0 \\\hline
+2 & q_\CommitIvk \cdot (a + 2^{130} - t_\mathbb{P} - a') = 0 \\\hline
+3 & q_\CommitIvk \cdot b_1 \cdot z_{a',13} = 0 \\\hline
+\end{array}
+$$
+
+### $\NullifierKey$ with $d_1 = 1 \implies \NullifierKey \geq 2^{254}$
+
+In these cases, we check that $\textsf{nk}_{0..=253} < t_\mathbb{P} < 2^{126}$:
+
+1. $d_1 = 1 \implies d_0 = 0.$
+
+   Since $d_1 = 1 \implies \NullifierKey_{0..=253} < 2^{126},$ we know that $\NullifierKey_{126..=253} = 0,$ and in particular $$d_0 := \NullifierKey_{245..=253} = 0.$$
+
+2. $d_1 = 1 \implies 0 \leq b_2 + 2^5 \cdot c < 2^{126}.$
+
+   To check that $0 \leq b_2 + 2^5 \cdot c < 2^{126}$, we use two constraints:
+
+    a) $0 \leq b_2 + 2^5 \cdot c < 2^{140}$. $b_2$ is already constrained individually to
+       be a $5$-bit value. $z_{c,13}$ is the index-13 running sum output by
+       $\SinsemillaHash(c).$ By constraining $$d_1 \cdot z_{c,13} = 0,$$ we constrain
+       $b_2 + 2^5 \cdot c < 2^{135} < 2^{140}.$
+
+    b) $0 \leq b_2 + 2^5 \cdot c + 2^{140} - t_\mathbb{P} < 2^{140}$. To check this, we
+       decompose ${b_2}c' = b_2 + 2^5 \cdot c + 2^{140} - t_\mathbb{P}$ into fourteen
+       10-bit words (little-endian) using a running sum $z_{{b_2}c'}$, looking up each
+       word in a $10$-bit lookup table. We then enforce in the custom gate that
+       $$d_1 \cdot z_{{b_2}c',14} = 0.$$
+
+$$
+\begin{array}{|c|l|}
+\hline
+\text{Degree} & \text{Constraint} \\\hline
+3 & q_\CommitIvk \cdot d_1 \cdot d_0 = 0 \\\hline
+3 & q_\CommitIvk \cdot d_1 \cdot z_{c,13} = 0 \\\hline
+2 & q_\CommitIvk \cdot (b_2 + c \cdot 2^5 + 2^{140} - t_\mathbb{P} - {b_2}c') = 0 \\\hline
+3 & q_\CommitIvk \cdot d_1 \cdot z_{{b_2}c',14} = 0 \\\hline
+\end{array}
+$$
+
+## Region layout
+
+The constraints controlled by the $q_\CommitIvk$ selector are arranged across all 10
+advice columns, requiring two rows.
+
+$$
+\begin{array}{|c|c|c|c|c|c|c|c|c|c|c}
+    &          &    &           &                 &               &     &     &     &     & q_\CommitIvk \\\hline
+a   & b        & c  & d         & \AuthSignPublic & \NullifierKey & b_0 & b_1 & b_2 & d_0 & 0 \\\hline
+d_1 & z_{a,13} & a' & z_{a',13} & z_{c,13}        & {b_2}c'       & z_{{b_2}c',14}  & & & & 1 \\\hline
+\end{array}
+$$
```

Comment:
I would prefer this region layout:
```suggestion
The constraints controlled by the $q_\CommitIvk$ selector are arranged across 9
advice columns, requiring two rows.

$$
\begin{array}{|c|c|c|c|c|c|c|c|c|c}
                &   &   &     &     &     &          &         &                & q_\CommitIvk \\\hline
\AuthSignPublic & a & b & b_0 & b_1 & b_2 & z_{a,13} & a'      & z_{a',13}      & 0 \\\hline
\NullifierKey   & c & d & d_0 & d_1 &     & z_{c,13} & {b_2}c' & z_{{b_2}c',14} & 1 \\\hline
\end{array}
$$
```

Reply:
This suggested layout has been implemented in https://github.com/zcash/orchard/pull/101/commits/e219adc923d245272f489fc905f4de99d1045c92.